From 57a1abd173203871f5afa94cd9b4b296783653ed Mon Sep 17 00:00:00 2001
From: Simon Que <sque@ti.com>
Date: Fri, 31 Jul 2009 21:30:03 -0500
Subject: [PATCH] Subject: SYSLINK fixes for shutdown stage

This patch fixes the setup/destroy of several modules to make
the cleanup stage run to completion.

Signed-off-by: Simon Que <sque@ti.com>
---
 drivers/dsp/syslink/multicore_ipc/messageq.c     |    2 +-
 drivers/dsp/syslink/multicore_ipc/nameserver.c   |   12 ++++++------
 drivers/dsp/syslink/multicore_ipc/sharedregion.c |    2 +-
 drivers/dsp/syslink/multicore_ipc/sysmgr.c       |    7 +++++--
 4 files changed, 13 insertions(+), 10 deletions(-)
 mode change 100644 => 100755 drivers/dsp/syslink/multicore_ipc/messageq.c
 mode change 100644 => 100755 drivers/dsp/syslink/multicore_ipc/sharedregion.c
 mode change 100644 => 100755 drivers/dsp/syslink/multicore_ipc/sysmgr.c

diff --git a/drivers/dsp/syslink/multicore_ipc/messageq.c b/drivers/dsp/syslink/multicore_ipc/messageq.c
old mode 100644
new mode 100755
index 5392400..00335fb
--- a/drivers/dsp/syslink/multicore_ipc/messageq.c
+++ b/drivers/dsp/syslink/multicore_ipc/messageq.c
@@ -445,7 +445,7 @@ int messageq_destroy(void)
 
 	if (!(atomic_dec_return(&messageq_state.ref_count)
 					== MESSAGEQ_MAKE_MAGICSTAMP(0))) {
-		status = -EBUSY;
+		status = 1;
 		goto exit;
 	}
 
diff --git a/drivers/dsp/syslink/multicore_ipc/nameserver.c b/drivers/dsp/syslink/multicore_ipc/nameserver.c
index 1a57ee4..8e3a5af 100644
--- a/drivers/dsp/syslink/multicore_ipc/nameserver.c
+++ b/drivers/dsp/syslink/multicore_ipc/nameserver.c
@@ -285,6 +285,12 @@ int nameserver_destroy(void)
 		goto exit;
 	}
 
+	if (!(atomic_dec_return(&nameserver_state.ref_count)
+					== NAMESERVER_MAKE_MAGICSTAMP(0))) {
+		retval = 1;
+		goto exit;
+	}
+
 	if (WARN_ON(nameserver_state.list_lock == NULL)) {
 		retval = -ENODEV;
 		goto exit;
@@ -296,12 +302,6 @@ int nameserver_destroy(void)
 		goto exit;
 	}
 
-	if (!(atomic_dec_return(&nameserver_state.ref_count)
-					== NAMESERVER_MAKE_MAGICSTAMP(0))) {
-		retval = -EBUSY;
-		goto exit;
-	}
-
 	retval = mutex_lock_interruptible(nameserver_state.list_lock);
 	if (retval)
 		goto exit;
diff --git a/drivers/dsp/syslink/multicore_ipc/sharedregion.c b/drivers/dsp/syslink/multicore_ipc/sharedregion.c
old mode 100644
new mode 100755
index ec2ed0e..1738388
--- a/drivers/dsp/syslink/multicore_ipc/sharedregion.c
+++ b/drivers/dsp/syslink/multicore_ipc/sharedregion.c
@@ -185,7 +185,7 @@ int sharedregion_destroy(void)
 
 	if (!(atomic_dec_return(&sharedregion_state.ref_count)
 					== SHAREDREGION_MAKE_MAGICSTAMP(0))) {
-		retval = -EBUSY; /* Syslink is not handling this on 2.0.0.06 */
+		retval = 1; /* Syslink is not handling this on 2.0.0.06 */
 		goto error;
 	}
 
diff --git a/drivers/dsp/syslink/multicore_ipc/sysmgr.c b/drivers/dsp/syslink/multicore_ipc/sysmgr.c
old mode 100644
new mode 100755
index 495c5aa..a381895
--- a/drivers/dsp/syslink/multicore_ipc/sysmgr.c
+++ b/drivers/dsp/syslink/multicore_ipc/sysmgr.c
@@ -578,8 +578,11 @@ s32 sysmgr_destroy(void)
 	}
 
 	if (atomic_dec_return(&sysmgr_state.ref_count)
-			== SYSMGR_MAKE_MAGICSTAMP(0))
-			goto exit;
+			!= SYSMGR_MAKE_MAGICSTAMP(0))
+    {
+		status = 1;
+		goto exit;
+    }
 
 	/* Finalize Platform module*/
 	if (sysmgr_state.platform_init_flag == true) {
-- 
1.5.4.3

